frontend/src/lib/components/Users/UserProjects.svelte

```diff
@@ -0,0 +1,59 @@
+<script  context="module" lang="ts">
+  // We define the Project type that we'll want in here, and export it so that callers can know they're passing in the right type
+  export type Project = {
+    id: string
+    name: string
+    code: string
+    role: ProjectRole
+  };
+</script>
+
+<script lang="ts">
+  import { ProjectRole } from '$lib/gql/types';
+  import { writable } from 'svelte/store';
+  import Badge from '../Badges/Badge.svelte';
+  import { onMount } from 'svelte';
+
+  export let projects: Project[] = [];
+  export let selectedProjects: string[] = [];
+
+  $: allSelected = projects && selectedProjects && selectedProjects.length === projects.length;
+
+  function handleSelectAllClick(): void {
+    if (!selectedProjects || !projects) return;
+    if (allSelected) {
+      selectedProjects = [];
+    } else {
+      selectedProjects = [...projects.map(proj => proj.code)];
+    }
+  }
+
+  function isManager(proj: Project): boolean {
+    return proj.role === ProjectRole.Manager;
+  }
+
+  // Projects managed by the given user come pre-checked, to save time in typical uses of this component
+  $: projectsStore = writable(projects);
+  onMount(() => projectsStore.subscribe(projects => {
```

Comment:
is there a simpler way to do this? maybe with something like `$: selectManagerProjects(projects)` it will only rerun when the input changes, but it will modify the selected projects?

Reply:
The simpler way breaks the "Select All" checkbox, because Svelte sees `selectedProjects` being modified and re-runs the `$:` block, resetting it to be only a list of projects managed by the user. Yes, even if I do it this way:

```ts
  function selectManagerProjects(projects: Project[]): void {
    selectedProjects = [... projects.filter(isManager).map(proj => proj.code)]
  }
  $: selectManagerProjects(projects);
```

Interestingly, it only resets the selection the first time you check or uncheck one of the boxes. The second time you try to check a box, the `$:` block doesn't re-run. So I think we do need to go with stores here, in order to be explicit about what we depend on (ONLY the projects list, don't re-run if checking a box ends up changing `selectedProjects`).